client/client.go

```diff
@@ -0,0 +1,227 @@
+		fqdn:     fqdn,
+		proxyUrl: proxyURL,
+	}
+	return c, nil
+}
+
+func (c *Coordinator) Start(ctx context.Context) {
+	c.loop(ctx)
+}
+
+func (c *Coordinator) handleErr(request *http.Request, err error) {
+	c.logger.Error("Coordinator error", "error", err)
+	scrapeErrorCounter.Inc()
+	resp := &http.Response{
+		StatusCode: http.StatusInternalServerError,
+		Body:       io.NopCloser(strings.NewReader(err.Error())),
+		Header:     http.Header{},
+	}
+	if err = c.doPush(resp, request); err != nil {
+		pushErrorCounter.Inc()
+		c.logger.Warn("Failed to push failed scrape response:", "err", err)
+		return
+	}
+	c.logger.Info("Pushed failed scrape response")
+}
+
+func (c *Coordinator) doScrape(request *http.Request) {
+	logger := c.logger.With("scrape_id", request.Header.Get("id"))
+	timeout, err := util.GetHeaderTimeout(request.Header)
+	if err != nil {
+		c.handleErr(request, err)
+		return
+	}
+	ctx, cancel := context.WithTimeout(request.Context(), timeout)
+	defer cancel()
+	request = request.WithContext(ctx)
+	// We cannot handle https requests at the proxy, as we would only
+	// see a CONNECT, so use a URL parameter to trigger it.
+	params := request.URL.Query()
+	if params.Get("_scheme") == "https" {
+		request.URL.Scheme = "https"
+		params.Del("_scheme")
+		request.URL.RawQuery = params.Encode()
+	}
+
+	if request.URL.Hostname() != c.fqdn {
+		c.handleErr(request, errors.New("scrape target doesn't match client fqdn"))
+		return
+	}
+
+	scrapeResp, err := c.client.Do(request)
+	if err != nil {
+		c.handleErr(request, fmt.Errorf("failed to scrape %s: %w", request.URL.String(), err))
+		return
+	}
+	logger.Info("Retrieved scrape response")
+	if err = c.doPush(scrapeResp, request); err != nil {
+		pushErrorCounter.Inc()
+		logger.Warn("Failed to push scrape response:", "err", err)
+		return
+	}
+	logger.Info("Pushed scrape result")
+}
+
+// Report the result of the scrape back up to the proxy.
+func (c *Coordinator) doPush(resp *http.Response, origRequest *http.Request) error {
+	resp.Header.Set("id", origRequest.Header.Get("id")) // Link the request and response
+	// Remaining scrape deadline.
+	deadline, _ := origRequest.Context().Deadline()
+	resp.Header.Set("X-Prometheus-Scrape-Timeout", fmt.Sprintf("%f", float64(time.Until(deadline))/1e9))
+
+	base, err := url.Parse(c.proxyUrl)
+	if err != nil {
+		return err
+	}
+	u, err := url.Parse("push")
+	if err != nil {
+		return err
+	}
+	url := base.ResolveReference(u)
+
+	buf := &bytes.Buffer{}
+	//nolint:errcheck // https://github.com/prometheus-community/PushProx/issues/111
+	resp.Write(buf)
+	request := &http.Request{
+		Method:        "POST",
+		URL:           url,
+		Body:          io.NopCloser(buf),
+		ContentLength: int64(buf.Len()),
+	}
+	request = request.WithContext(origRequest.Context())
+	if _, err = c.client.Do(request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Coordinator) doPoll(ctx context.Context) error {
+	base, err := url.Parse(c.proxyUrl)
+	if err != nil {
```

Comment:
you have the same parse and error check in `doPush` but without logging error, why does it log and error here but not there?

if they should both log or not log an error, maybe make it a function as its a 4x repeating code block